src/vs/base/browser/ui/selectBox/selectBox.ts

```diff
@@ -5,56 +5,190 @@
+	selectOptionHoverForeground: Color.fromHex('#ffffff'),
+	selectOptionHoverBackground: Color.fromHex('#2A2D2E'),
+	selectOptionCheckedOutline: Color.fromHex('#F38518'),
+	selectOptionHoverOutline: Color.fromHex('#F38518')
 };
 
 export interface ISelectData {
 	selected: string;
 	index: number;
 }
 
-export class SelectBox extends Widget {
+export class SelectBox extends Widget implements IDelegate<ISelectOptionItem> {
+
+	private static SELECT_DROPDOWN_BOTTOM_MARGIN = 10;
 
+	private _useNativeSelect: boolean;
 	private selectElement: HTMLSelectElement;
 	private options: string[];
 	private selected: number;
+	private disabledOptionIndex: number;
 	private _onDidSelect: Emitter<ISelectData>;
 	private toDispose: IDisposable[];
-	private selectBackground: Color;
-	private selectForeground: Color;
-	private selectBorder: Color;
+	private styles: ISelectBoxStyles;
+	private listRenderer: SelectListRenderer;
+	private contextViewProvider: IContextViewProvider;
+	private selectDropDownContainer: HTMLElement;
+	private styleElement: HTMLStyleElement;
+	private selectList: List<ISelectOptionItem>;
+	private selectDropDownListContainer: HTMLElement;
+	private widthControlElement: HTMLElement;
+	// private widthControDivElement: HTMLElement;
 
-	constructor(options: string[], selected: number, styles: ISelectBoxStyles = deepClone(defaultStyles)) {
+	constructor(options: string[], selected: number, contextViewProvider: IContextViewProvider, styles: ISelectBoxStyles = deepClone(defaultStyles)) {
 		super();
 
+		// Hardcode native select switch based on platform
+		switch (process.platform) {
+			case 'darwin':
+				this._useNativeSelect = true;
+				break;
+			case 'win32':
+			default:
+				this._useNativeSelect = false;
+				break;
+		}
+
 		this.selectElement = document.createElement('select');
 		this.selectElement.className = 'select-box';
 
-		this.setOptions(options, selected);
 		this.toDispose = [];
 		this._onDidSelect = new Emitter<ISelectData>();
 
-		this.selectBackground = styles.selectBackground;
-		this.selectForeground = styles.selectForeground;
-		this.selectBorder = styles.selectBorder;
+		this.styles = styles;
+		mixin(this.styles, defaultStyles, false);
+
+		this.registerNativeSelectListeners();
+
+		if (!this._useNativeSelect) {
+			this.constructSelectDropDown(contextViewProvider);
+			this.registerSelectDropDownListeners();
+		}
+
+		this.setOptions(options, selected);
+	}
+
+	// IDelegate - List renderer
+
+	getHeight(): number {
+		return 18;
+	}
+
+	getTemplateId(): string {
+		return SELECT_OPTION_ENTRY_TEMPLATE_ID;
+	}
+
+	private constructSelectDropDown(contextViewProvider: IContextViewProvider) {
+
+		// SetUp ContextView container to hold select Dropdown
+		this.contextViewProvider = contextViewProvider;
+		this.selectDropDownContainer = dom.$('.select-dropdown-container');
+
+		// Setup list for drop-down select
+		this.createSelectList(this.selectDropDownContainer);
+
+		// TODO:cleidigh - find better width control
```

Comment:
@cleidigh can you fix or remove this TODO?

Reply:
Removed